# HG changeset patch
# Parent c9c3e02c44f94eaf7860a8fbee5b536ef2ccaaa6
Fixed the handling of vettical row changing using the keyboard in wxDVC. It was
possible to get to a an invalid row when items were deleted because the 'current'
row was not properly reset.

diff -r c9c3e02c44f9 -r bf5a1ff970b0 src/generic/datavgen.cpp
--- a/src/generic/datavgen.cpp	Mon Aug 20 09:49:03 2012 +0200
+++ b/src/generic/datavgen.cpp	Mon Sep 24 09:50:48 2012 +0200
@@ -602,7 +602,7 @@
     void OnPaint( wxPaintEvent &event );
     void OnCharHook( wxKeyEvent &event );
     void OnChar( wxKeyEvent &event );
-    void OnVerticalNavigation(unsigned int newCurrent, const wxKeyEvent& event);
+    void OnVerticalNavigation(int steps, const wxKeyEvent& event);
     void OnLeftKey();
     void OnRightKey();
     void OnMouse( wxMouseEvent &event );
@@ -1387,7 +1387,7 @@
     m_currentCol = NULL;
     m_currentColSetByKeyboard = false;
     m_useCellFocus = false;
-    m_currentRow = 0;
+    m_currentRow = (unsigned)-1;
 
 #ifdef __WXMSW__
     // We would like to use the same line height that Explorer uses. This is
@@ -2390,8 +2390,11 @@
     }
 
     // Change the current row to the last row if the current exceed the max row number
-    if( m_currentRow > GetRowCount() )
+    if( m_currentRow >= GetRowCount() )
+    {
         ChangeCurrentRow(m_count - 1);
+        //RefreshRow(m_currentRow);
+    }
 
     GetOwner()->InvalidateColBestWidths();
     UpdateDisplay();
@@ -2464,6 +2467,7 @@
 {
     DestroyTree();
     m_selection.Clear();
+    m_currentRow = (unsigned)-1;
 
     SortPrepare();
     BuildTree( GetModel() );
@@ -2987,6 +2991,10 @@
 {
     wxASSERT( !IsVirtualList() );
 
+    // someone could have passed us the 'invalid row'
+    if(row == (unsigned)-1)
+        return 0;
+
     RowToTreeNodeJob job( row , -2, m_root );
     Walker( m_root , job );
     return job.GetResult();
@@ -3689,13 +3697,11 @@
             break;
 
         case WXK_UP:
-            if ( m_currentRow > 0 )
-                OnVerticalNavigation( m_currentRow - 1, event );
+            OnVerticalNavigation( -1, event );
             break;
 
         case WXK_DOWN:
-            if ( m_currentRow + 1 < GetRowCount() )
-                OnVerticalNavigation( m_currentRow + 1, event );
+            OnVerticalNavigation( +1, event );
             break;
         // Add the process for tree expanding/collapsing
         case WXK_LEFT:
@@ -3707,37 +3713,19 @@
             break;
 
         case WXK_END:
-        {
-            if (!IsEmpty())
-                OnVerticalNavigation( GetRowCount() - 1, event );
+            OnVerticalNavigation( +(int)GetRowCount(), event );
             break;
-        }
+
         case WXK_HOME:
-            if (!IsEmpty())
-                OnVerticalNavigation( 0, event );
+            OnVerticalNavigation( -(int)GetRowCount(), event );
             break;
 
         case WXK_PAGEUP:
-            {
-                int steps = pageSize - 1;
-                int index = m_currentRow - steps;
-                if (index < 0)
-                    index = 0;
-
-                OnVerticalNavigation( index, event );
-            }
+            OnVerticalNavigation( -(pageSize - 1), event );
             break;
 
         case WXK_PAGEDOWN:
-            {
-                int steps = pageSize - 1;
-                unsigned int index = m_currentRow + steps;
-                unsigned int count = GetRowCount();
-                if ( index >= count )
-                    index = count - 1;
-
-                OnVerticalNavigation( index, event );
-            }
+            OnVerticalNavigation( +(pageSize - 1), event );
             break;
 
         default:
@@ -3745,16 +3733,26 @@
     }
 }
 
-void wxDataViewMainWindow::OnVerticalNavigation(unsigned int newCurrent, const wxKeyEvent& event)
+void wxDataViewMainWindow::OnVerticalNavigation(int steps, const wxKeyEvent& event)
 {
-    wxCHECK_RET( newCurrent < GetRowCount(),
-                wxT("invalid item index in OnVerticalNavigation()") );
-
     // if there is no selection, we cannot move it anywhere
-    if (!HasCurrentRow())
+    if (!HasCurrentRow() || IsEmpty())
         return;
 
+    //TODO: change the type of row variables (e.g. m_currentRow) to int to make things easier
+
+    int newRow = (int)m_currentRow + steps;
+
+    // let's keep the new row inside the allowed range
+    if ( newRow < 0 )
+        newRow = 0;
+
+    const int rowCount = (int)GetRowCount();
+    if ( newRow >= rowCount )
+        newRow = rowCount - 1;
+
     unsigned int oldCurrent = m_currentRow;
+    unsigned int newCurrent = (unsigned int)newRow;
 
     // in single selection we just ignore Shift as we can't select several
     // items anyhow
@@ -3806,6 +3804,8 @@
     else
     {
         wxDataViewTreeNode* node = GetTreeNodeByRow(m_currentRow);
+        if(!node)
+            return;
 
         if ( TryAdvanceCurrentColumn(node, /*forward=*/false) )
             return;
@@ -3848,6 +3848,8 @@
     else
     {
         wxDataViewTreeNode* node = GetTreeNodeByRow(m_currentRow);
+        if(!node)
+            return;
 
         if ( node->HasChildren() )
         {
